File: stack/Car Fleet.cs
public class SolutionCarFleet
{
    public int CarFleet(int target, int[] position, int[] speed)
    {
        int n = position.Length;
        if (n == 0) return 0;

        (int pos, int spd)[] p_s = new (int, int)[n];
        for (int i = 0; i < n; ++i)
            p_s[i] = (position[i], speed[i]);

        Array.Sort(p_s, (a, b) => b.pos.CompareTo(a.pos));

        Stack<double> times = new Stack<double>();

        for (int i = 0; i < n; ++i)
        {
            double time = (double)(target - p_s[i].pos) / p_s[i].spd;
            if (times.Count == 0 || time > times.Peek())
                times.Push(time);
        }

        return times.Count;
    }
}

File: stack/Car Fleet.py
from typing import List

def carFleet(target: int, position: List[int], speed: List[int]) -> int:
        cars = sorted(zip(position, speed), reverse=True)

        stack = []
        for pos, spd in cars:
            time = (target - pos) / spd
            if not stack or time > stack[-1]:
                stack.append(time)

        return len(stack)

print(carFleet(12, [10, 8, 0, 5, 3], [2, 4, 1, 1, 3]))
